docs/architecture/adr-069-flexible-node-intitalization.md

```diff
@@ -0,0 +1,253 @@
+solution.
+
+### Generic Service Plugability
+
+We can imagine a system design that exports interfaces (in the Golang
+sense) for all components of the system, to permit runtime dependency
+injection of all components in the system so that users can compose
+tendermint nodes of arbitrary user supplied components.
+
+This is an interesting proposal, and may be interesting to persue, but
+essentially requires doing a lot more API design and increases the
+surface area of our API. While this is not a goal at the moment,
+eventually providing support for some kinds of plugability may be
+useful, so the current solution does not explicitly forclose the
+possibility of this alternative.
+
+### Abstract Dependency Based Startup and Shutdown
+
+The proposal on in this document simplifies and makes tendermint node
+initialization more abstract, but the system lacks a number of
+features which daemon/service initialization might provide, such as a
+dependency based system that allows the authors of services to control
+the initialization and shutdown order of components using an
+dependencies to control the ordering.
+
+Dependency based orderings make it possible to write components
+(reactors, etc.) with only limited awareness of other components or of
+node initialization, and is the state of the art for process
+initialization. However, this may be too abstract and complicated, and
+authors of components in the current implementation of tendermint
+*would* need to know about other components, so a dependency based
+system would be unhelpfully abstract at this stage.
+
+## Decisions
+
+- Provide a more flexible internal framework for initializing tendermint
+  nodes to make the initatilization process less hard-coded by the
+  implementation of the node objects.
+
+  - Reactors should not need to expose their interfaces *within* the
+	implementation of the node type, except in the context of some groups of
+	services.
+
+  - This refactoring should be entirely opaque to users.
+
+  - These node initialization changes should not require a
+	reevaluation of the `service.Service` or a generic initialization
+	orchestration framework.
+
+- If required for the 0.35 release, add an "unsafe" way to construct a
+  node service with user-supplied mempool, including making the
+  relevant interfaces and types external/exported.
+
+- Prioritize implementation of p2p-based statesync reactor to obviate
+  need for users to inject a custom state-sync provider.
+
+## Detailed Design
+
+The [current
+nodeImpl](https://github.com/tendermint/tendermint/blob/master/node/node.go#L47)
+includes direct references to the implementations of each of the
+reactors, which should be replaced by references to `service.Service`
+objects. This will require moving the construction of the [rpc
+service](https://github.com/tendermint/tendermint/blob/master/node/node.go#L771)
+into the constructor of
+[makeNode](https://github.com/tendermint/tendermint/blob/master/node/node.go#L126).
```

Comment:
I'm not completely sure what this looks like. Looking at where RPC is initialized currently, I'm definitely concerned about it. It is constructed and initialized within a method on the node. My strong preference is that, to whatever degree possible, the logic for configuring and starting this process be moved _into the RPC package_. 
I would generally advocate for deleting the `ConfigureRPC` and`startRPC` methods from node. How to create and start an RPC server is not its concern. It should instead be passed a constructed RPC server that is ready to run. 

(I would also advocate for changing the way rpc is constructed, but that may be a different discussions for a different time).

My preferred style for constructors of this type is definitely [the component graph](https://peter.bourgon.org/go-for-industrial-programming/#the-component-graph).

Reply:
Yeah, the `rpc` package is really where this knot falls into. I don't think in this change we can fully realize this goal, but I think we can do a bunch of refactoring to simplify the code.

The work to get the RPC layer to be independent of the `node` package and to be able to interface more abstractly with other components is something that I think we should agree to push out a bit (particularly if we need to get pluggable mempool into 0.35) because I think it's a *huge* amount of work, and while it would be better, I don't think the benefit would justify the work right now.